<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Differential expression using limma — limma • ezRNAseq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">ezRNAseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Differential expression using limma</h1>
    </div>

    
    <p>Differential expression using limma:
<ul>
<li> <b>lm_fit()</b>: Fit easily linear model for gene expression data using limma::lmFit().
</li>
<li> <b>lm_results()</b>: Extract results from lm_fit analysis. Compared to the function topTable(),
it performs automatically an independent filtering as described by Bourgon et al., PNAS 2010.
</li>
<li> <b>results_names()</b>: Returns the names of the estimated effects (coefficents) of the model
</li>
</ul></p>
    

    <pre><span class='fu'>lm_fit</span>(<span class='no'>eset</span>, <span class='no'>design</span>, <span class='no'>logged2</span>)

<span class='fu'>lm_results</span>(<span class='no'>object</span>, <span class='no'>contrast</span>, <span class='kw'>independentFiltering</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>alpha</span> <span class='kw'>=</span> <span class='fl'>0.05</span>,
  <span class='no'>filter</span>, <span class='no'>theta</span>, <span class='kw'>pAdjustMethod</span> <span class='kw'>=</span> <span class='st'>"BH"</span>, <span class='no'>...</span>)

<span class='fu'>results_names</span>(<span class='no'>object</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <dl class="dl-horizontal">
      <dt>eset</dt>
      <dd>an object of class ExpressionSet or a list with two components including data and samples.
eset must contain the information about samples.</dd>
      <dt>design</dt>
      <dd>a formula which expresses how the expression for each gene depend on the variables in the sample data.
e.g.: ~ 0 + group (design with multiple groups ),  ~ group + condition (design with multiple variables)
and ~ genotype + treatment + genotype:treatment (design with interactions).</dd>
      <dt>logged2</dt>
      <dd>logical value. Specify if the data are in log2 scale or not. Allowed values are TRUE or FALSE.</dd>
      <dt>object</dt>
      <dd>an object of class &quot;lm_fit&quot; generated from the function lm_fit()</dd>
      <dt>contrast</dt>
      <dd>specifies what comparison to extract from the object to build a results table.
Possible values include:
<ul>
<li> a character vector with three elements:
the name of a factor in the design formula, the name of group 1 and group 2 to be compared;
e.g.: c(&quot;group&quot;, &quot;grp1&quot;, &quot;grp2&quot;). &quot;grp1&quot; is used as numerator and &quot;grp2&quot; as denominator for
computing the fold change.
</li>
<li> a list of 2 character vectors to be compared. These names should be elements of lm_results_names(object).
e.g.: list(c(&quot;grp1&quot;,&quot;grp2&quot;), c(&quot;grp3&quot;, &quot;group4&quot;))
</li>
</ul>
</dd>
      <dt>independentFiltering</dt>
      <dd>logical, whether independent filtering should be applied automatically</dd>
      <dt>alpha</dt>
      <dd>the significance cutoff used for optimizing the independent filtering (by default 0.1).</dd>
      <dt>filter</dt>
      <dd>the vector of filter statistics over which the independent filtering will be optimized.</dd>
      <dt>theta</dt>
      <dd>the quantiles at which to assess the number of rejections from independent filtering</dd>
      <dt>pAdjustMethod</dt>
      <dd>method used to adjust the p-values for multiple testing.
Allowed values include &quot;none&quot;, &quot;BH&quot;, &quot;BY&quot; and &quot;holm&quot; (See ?p.adjust).</dd>
      <dt>...</dt>
      <dd>other arguments to be passed to the function limma::topTable</dd>
    </dl>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><ul>
<li> <b>lm_fit()</b>: Returns an object of class &quot;MArrayLM&quot; (see limma:?lmFit) and &quot;lm_fit&quot;.
Comprared to lmFit(), the object returned by lm_fit() contains supplementary variables named:
<ul>
<li> design_formula: which holds the design formula.
</li>
<li> samples: annotation of each samples as provided in eset.
</li>
<li> data: unlogged gene expression data
</li>
</ul></p>
    <p></li>
<li> <b>lm_results()</b>: Returns an object of class &quot;DE_Results&quot; which is a simple subclass of DataFrame (in S4Vectors)
containing log2 fold changes, base means across samples, t-test statistics, p-values
adjusted p-values. The function plot_MA() can be used to draw the results. (See ?plot_MA).
</li>
<li> <b>results_names()</b>: Returns the names of the estimated effects (coefficents) of the model.
</li>
</ul></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># Load and prepare data</span>
<span class='fu'>data</span>(<span class='no'>ALL_bcrneg</span>)
<span class='no'>ALL_bcrneg</span>$<span class='no'>mol.biol</span> <span class='kw'>&lt;-</span> <span class='fu'>factor</span>(<span class='fu'>make.names</span>(<span class='no'>ALL_bcrneg</span>$<span class='no'>mol.biol</span>),
<span class='kw'>levels</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"NEG"</span>, <span class='st'>"BCR.ABL"</span>))

<span class='co'># Fit linear model</span>
<span class='no'>fit</span> <span class='kw'>&lt;-</span> <span class='fu'>lm_fit</span>(<span class='no'>ALL_bcrneg</span>, <span class='kw'>design</span> <span class='kw'>=</span> ~<span class='fl'>0</span>+<span class='no'>mol.biol</span>, <span class='kw'>logged2</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># result names</span>
<span class='fu'>results_names</span>(<span class='no'>fit</span>)</div><div class='output co'>#&gt; [1] &quot;mol.biolNEG&quot;     &quot;mol.biolBCR.ABL&quot;</div><div class='input'>
<span class='co'># Results of differential expression</span>
<span class='no'>res</span> <span class='kw'>&lt;-</span> <span class='fu'>lm_results</span>(<span class='no'>fit</span>)

<span class='co'># Plot MA</span>
<span class='fu'><a href='plot_MA.html'>plot_MA</a></span>(<span class='no'>res</span>)</div><img src='limma-4.png' alt='' width='540' height='400' /><div class='input'>
<span class='co'># Independent filtering threshold and rejection plot</span>
 <span class='fu'>attr</span>(<span class='no'>res</span>,<span class='st'>"filterThreshold"</span>)</div><div class='output co'>#&gt;      10% 
#&gt; 10.18845 </div><div class='input'>
 <span class='fu'><a href='plot.html'>plot</a></span>(<span class='fu'>attr</span>(<span class='no'>res</span>,<span class='st'>"filterNumRej"</span>),<span class='kw'>type</span><span class='kw'>=</span><span class='st'>"b"</span>,
 <span class='kw'>xlab</span><span class='kw'>=</span><span class='st'>"quantiles of 'baseMean'"</span>,
 <span class='kw'>ylab</span><span class='kw'>=</span><span class='st'>"number of rejections"</span>)</div><img src='limma-8.png' alt='' width='540' height='400' /><div class='input'>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Alboukadel Kassambara.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
