<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>MA-plot from means and log fold changes — plot_MA • ezRNAseq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">ezRNAseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>MA-plot from means and log fold changes</h1>
    </div>

    
    <p>Make MA-plot which is a scatter plot of log2 fold changes (on the y-axis)
versus the mean expression signal (on the x-axis).</p>
    

    <pre><span class='fu'>plot_MA</span>(<span class='no'>x</span>, <span class='kw'>fdr</span> <span class='kw'>=</span> <span class='fl'>0.05</span>, <span class='kw'>fc</span> <span class='kw'>=</span> <span class='fl'>1.5</span>, <span class='kw'>detection_call</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>ylim</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>colNonSig</span> <span class='kw'>=</span> <span class='st'>"darkgray"</span>, <span class='kw'>colUp</span> <span class='kw'>=</span> <span class='st'>"#BC002A"</span>, <span class='kw'>colDown</span> <span class='kw'>=</span> <span class='st'>"#16614E"</span>,
  <span class='kw'>colLine</span> <span class='kw'>=</span> <span class='st'>"black"</span>, <span class='kw'>log</span> <span class='kw'>=</span> <span class='st'>"x"</span>, <span class='kw'>cex</span> <span class='kw'>=</span> <span class='fl'>0.45</span>, <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"Mean expression"</span>,
  <span class='kw'>ylab</span> <span class='kw'>=</span> <span class='st'>"Log2 fold change"</span>, <span class='kw'>addlegend</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>highlight</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='no'>...</span>)

<span class='co'># S3 method for DESeqResults</span>
<span class='fu'><a href='plot.html'>plot</a></span>(<span class='no'>x</span>, <span class='no'>...</span>)

<span class='co'># S3 method for DE_Results</span>
<span class='fu'><a href='plot.html'>plot</a></span>(<span class='no'>x</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <dl class="dl-horizontal">
      <dt>x</dt>
      <dd>an object of class DESeqResults, get_diff, DE_Results, matrix or data frame containing the columns
baseMean, log2FoldChange, and padj. Rows are genes.
<ul>
<li> baseMean: the mean expression of genes in the two groups.
</li>
<li> log2FoldChange: the log2 fold changes of group 2 compared to group 1
</li>
<li> padj: the adjusted p-value of the used statiscal test.
</li>
</ul>
</dd>
      <dt>fdr</dt>
      <dd>Accepted false discovery rate for considering genes as differentially expressed.</dd>
      <dt>fc</dt>
      <dd>the fold change threshold. Only genes with a fold change &gt;= fc and padj &lt;= fdr are
considered as significantly differentially expressed.</dd>
      <dt>detection_call</dt>
      <dd>a numeric vector with length = nrow(x),
specifying if the genes is expressed or not. Default is NULL. If detection_call column is available in x,
it will be used.</dd>
      <dt>ylim</dt>
      <dd>The limits for the y-axis.
If missing, an attempt is made to choose a sensible value.
Dots exceeding the limits will be displayed as triangles at the limits, pointing outwards.</dd>
      <dt>colNonSig, colUp, colDown</dt>
      <dd>colors to be used for non-significant,
up-regulated and down-regulated data points, respectively.</dd>
      <dt>colLine</dt>
      <dd>colour to use for the horizontal (y=0) line.</dd>
      <dt>log</dt>
      <dd>which axis/axes should be logarithmic; will be passed to plot.</dd>
      <dt>cex</dt>
      <dd>point size</dd>
      <dt>xlab, ylab</dt>
      <dd>x and y axis labels</dd>
      <dt>addlegend</dt>
      <dd>a logical value. If TRUE a legend is added to the plot.</dd>
      <dt>highlight</dt>
      <dd>highlight top genes. Default value is 10 for top 10 genes.</dd>
      <dt>...</dt>
      <dd>other parameters to be passed to the function plot</dd>
    </dl>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>returns a base plot</p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Alboukadel Kassambara.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
