<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>RNAseq Workflow — rnaseq_workflow • ezRNAseq</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>

<!-- mathjax -->
<script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">ezRNAseq</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>RNAseq Workflow</h1>
    </div>

    
    <p>RNAseq Workflow. Alignment with STAR and read counting with R/Bioconductor.</p>
    

    <pre><span class='fu'>rnaseq_workflow</span>(<span class='kw'>data_dir</span> <span class='kw'>=</span> <span class='fu'>getwd</span>(), <span class='kw'>samples.annotation</span> <span class='kw'>=</span> <span class='st'>"samples.txt"</span>,
  <span class='kw'>pairedEnd</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>fastq.gz</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>star.index</span> <span class='kw'>=</span> <span class='st'>"/eqmoreaux/genomes/Homo_sapiens/Ensembl/GRCh37/Sequence/StarIndex/"</span>,
  <span class='kw'>gtf</span> <span class='kw'>=</span> <span class='st'>"/eqmoreaux/genomes/Homo_sapiens/Ensembl/GRCh37/Annotation/Genes/genes.gtf"</span>,
  <span class='kw'>result.dir</span> <span class='kw'>=</span> <span class='fu'>getwd</span>(), <span class='kw'>keep</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"name_sorted_bam"</span>),
  <span class='kw'>ignore.strand</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>count.mode</span> <span class='kw'>=</span> <span class='st'>"Union"</span>, <span class='kw'>thread</span> <span class='kw'>=</span> <span class='fl'>10</span>,
  <span class='kw'>data.analyst</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='kw'>name</span> <span class='kw'>=</span> <span class='st'>""</span>, <span class='kw'>email</span> <span class='kw'>=</span> <span class='st'>""</span>), <span class='kw'>data.author</span> <span class='kw'>=</span> <span class='fu'>list</span>(<span class='kw'>name</span> <span class='kw'>=</span> <span class='st'>""</span>,
  <span class='kw'>email</span> <span class='kw'>=</span> <span class='st'>""</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <dl class="dl-horizontal">
      <dt>data_dir</dt>
      <dd>the directory containing the data. It should contains a subdirectory called &quot;FASTQ&quot; which
contains fastq files. Default value is the working directory</dd>
      <dt>samples.annotation</dt>
      <dd>the path to &quot;samples.txt&quot; tabulation file containing sample annotations.
This file should contain at least 3 columns:
<ul>
<li> name: corresponding to sample name. It will be used to report the result
</li>
<li> fastq1: the first fastq file name for paired-end sequencing
</li>
<li> fastq2: the second fastq file name for paired-end sequencing.
This is optional if single-end sequencing is used
</li>
</ul>
</dd>
      <dt>pairedEnd</dt>
      <dd>specify if the data are paired-end sequencing data. Default is TRUE.</dd>
      <dt>fastq.gz</dt>
      <dd>specify if the FASTQ files are compressed. Default is TRUE.</dd>
      <dt>star.index</dt>
      <dd>reference genome for STAR.</dd>
      <dt>gtf</dt>
      <dd>gene annotation file.</dd>
      <dt>result.dir</dt>
      <dd>The absolute path of result directory.</dd>
      <dt>keep</dt>
      <dd>a character vector specifying the file to be kept.
Allowed values include c(&quot;name_sorted_bam&quot;, &quot;chr_sorted_bam&quot;, &quot;sam&quot; ).</dd>
      <dt>ignore.strand</dt>
      <dd>A logical indicating if strand should be considered when
matching. if TRUE, allows minus strand reads to count in plus strand genes,
and vice versa. Should be FALSE for stranded RNAsequencing.</dd>
      <dt>count.mode</dt>
      <dd>counting mode. Read ?GenomicAlignments::summarizeOverlaps. Default
value is &quot;Union&quot;. Reads that overlap any portion of exactly one feature are
counted. Reads that overlap multiple features are discarded.</dd>
      <dt>thread</dt>
      <dd>Number of threads to be used. This depends to the available computer ressources.</dd>
      <dt>data.analyst</dt>
      <dd>a list containing the name and the email of the data analyst.</dd>
      <dt>data.author</dt>
      <dd>a list containing the name and the email of the data author.</dd>
    </dl>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Can create three subdirectories: 
<ul>
<li> SAM: containing the output of STAR alignment program
</li>
<li> BAM: containing unsorted and sorted BAM files, and BAM index file.
BAM comtent is sorted by read names (*_name_sorted.bam) or by chromosome (*_sorted.bam).
The index files are of form *_sorted.bam.bai. *_name_sorted.bam files are used for read counting
using hseq-count or R. *_sorted.bam and *_sorted.bam.bai files can be used for IGV visualization.
</li>
<li> COUNT: containing read counting results. It contains the following files:
1. se.RDATA: containing an object &quot;se&quot; which is an object of class SummarizedExperiment
</li>
</ul></p>
    

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
          </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Alboukadel Kassambara.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
